holoviews/core/data.py

```diff
@@ -1308,9 +1334,288 @@ def aggregate(cls, columns, kdims, function, **kwargs):
 
 
 
+class GridColumns(DictColumns):
+    """
+    Interface for simple dictionary-based columns format. The dictionary
+    keys correspond to the column (i.e dimension) names and the values
+    are collections representing the values in that column.
+    """
+
+    types = (dict, OrderedDict, cyODict)
+
+    datatype = 'grid'
+
+    @classmethod
+    def reshape(cls, eltype, data, kdims, vdims):
+        if kdims is None:
+            kdims = eltype.kdims
+        if vdims is None:
+            vdims = eltype.vdims
+
+        if not vdims:
+            raise ValueError('GridColumns interface requires at least '
+                             'one value dimension.')
+
+        dimensions = [d.name if isinstance(d, Dimension) else
+                      d for d in kdims + vdims]
+        if isinstance(data, tuple):
+            data = {d: v for d, v in zip(dimensions, data)}
+        elif not isinstance(data, dict):
+            raise ValueError('GridColumns must be instantiated as a '
+                             'dictionary or tuple')
+
+        if 'vdims' in data:
+            vdim_array = data.pop('vdims')
+            for i, vdim in enumerate(vdims):
+                name = vdim.name if isinstance(vdim, Dimension) else vdim
+                data[name] = vdim_array[..., i]
+
+        for dim in kdims+vdims:
+            name = dim.name if isinstance(dim, Dimension) else dim
+            if name not in data:
+                raise ValueError("Values for dimension %s not found" % dim)
+            if not isinstance(data[name], np.ndarray):
+                data[name] = np.array(data[name])
+
+        kdim_names = [d.name if isinstance(d, Dimension) else d for d in kdims]
+        vdim_names = [d.name if isinstance(d, Dimension) else d for d in vdims]
+        expected = [len(data[kd]) for kd in kdim_names]
+        for vdim in vdim_names:
+            shape = data[vdim].shape
+            if shape != tuple(expected):
+                raise ValueError('Key dimension values and value array %s'
+                                 'shape do not match. Expected shape %s, '
+                                 'actual shape: %s' % (expected, vdim, shape))
+        return data, kdims, vdims
+
+
+    @classmethod
+    def validate(cls, columns):
+        DataColumns.validate(columns)
+
+
+    @classmethod
+    def dimension_type(cls, columns, dim):
+        if dim in columns.kdims:
+            arr = columns.data[dim.name]
+        elif dim in columns.vdims:
+            arr = columns.data[dim.name]
+        else:
+            return None
+        return arr.dtype.type
+
+
+    @classmethod
+    def shape(cls, columns):
+        return cls.length(columns), len(columns.dimensions()),
+
+
+    @classmethod
+    def length(cls, columns):
+        return np.product([len(columns.data[d.name]) for d in columns.kdims])
+
+
+    @classmethod
+    def values(cls, columns, dim, expanded=True, flat=True):
+        if dim in columns.kdims:
```

Comment:
The docstring will need to explain that the cartesian product is only applied to key dimensions.